### ISSUE TITLE
Help.I write a deepid network using mxnet,but it is always getting the maximum Train-accuracy of 0.405538

### ISSUE THREAD

**Issue by dushoufu** [dmlc/mxnet#2246]
I am tasked with transfering deepid using caffe to that using mxnet. All parameters come from a caffe prototxt. But how I train the network,the maximum Train-accuracy is only up  to like 0.40,while caffe can reach at least 80%. I think possibly my mxnet network has something wrong. Could some experts help check my programm and give some advice? Thank very much.

In this program, I changed a source code that belongs to mxnet original sources,which is python\mxnet\initializer.py. My aim is to initialize every layers with different initial weight and bias.
In initializer.py I add a class,as following:

```python
class Custom(Initializer):
    """Initialize the weight and bias with custom requirements

    Parameters
    ----------
    {"layername":{"initialiseMethod":(para1,para2)}}
    initialiseMethod:
    for weight: "normal","uniform","orthogonal","xavier"
    for bias:"costant"
    dict : , optional
        Standard deviation for gaussian distribution.
    """
    weightMethods=["normal","uniform","orthogonal","xavier"]
    biasMethods=["costant"]
    
    def __init__(self, idict):
        self.dict=idict

    def _init_weight(self, name, arr):
        if name in self.dict.keys():
        	dictPara=self.dict[name]
        	for(k, v) in dictPara.items():
        		if k.lower()=="normal":
        			random.normal(0, v, out=arr)
        		elif k.lower()=="uniform":
        			raise NotImplementedError("Not support at the moment")
        		elif k.lower()=="orthogonal":
        			raise NotImplementedError("Not support at the moment")
        		elif k.lower()=="xavier":
        			xa = Xavier(v[0],v[1],v[2])
        			xa(name,arr)
        else:
        	raise NotImplementedError("Not support")
        
    def _init_bias(self, name, arr):
        if name in self.dict.keys():
        	dictPara=self.dict[name]
        	for(k, v) in dictPara.items():
        		if k.lower()=="costant":
        			arr[:]=v
        else
        	raise NotImplementedError("Not support")
```

my data.py is:
```python
# pylint: skip-file
""" data iterator for deepid """
import sys
import os
import mxnet as mx

def deepid_iterator(batch_size, input_shape):
    """return train and val iterators for deepid"""
    
    train = mx.io.ImageRecordIter(
        path_imgrec = "./train.rec",
        data_shape  = input_shape,
        batch_size  = batch_size,
        rand_crop   = True,
        rand_mirror = True,
        shuffle     = True,
				mirror      = True,
#        scale       = 0.00390625,
        round_batch = True)
    val = mx.io.ImageRecordIter(
        path_imgrec = "./test.rec",
        data_shape  = input_shape,
        batch_size  = batch_size,
        rand_crop   = False,
        rand_mirror = False,
        shuffle     = False,
				mirror      = False,
#        scale       = 0.00390625,
        round_batch = False)

    return (train, val)

my deepid_py is:
# pylint: skip-file
from data import deepid_iterator
import mxnet as mx
import logging
import datetime

#logging the network running
prestr=datetime.datetime.now().strftime("%Y%m%d%H%M")
outfile=prestr+".txt"

## define deepid
input_data = mx.symbol.Variable(name="data")

# stage 1
conv1 = mx.symbol.Convolution(data=input_data, kernel=(4, 4), stride=(1, 1), num_filter=20,workspace=1024)
relu1 = mx.symbol.Activation(data=conv1, act_type="relu")
pool1 = mx.symbol.Pooling(
    data=relu1, pool_type="max", kernel=(2, 2), stride=(2,2))

#arg_shape, out_shape, aux_shape = conv1.infer_shape(date=(256,1,100,100))

# stage 2
conv2 = mx.symbol.Convolution(data=pool1, kernel=(3, 3), num_group=2, num_filter=40,workspace=1024)
relu2 = mx.symbol.Activation(data=conv2, act_type="relu")
pool2 = mx.symbol.Pooling(data=relu2, kernel=(2, 2), stride=(2, 2), pool_type="max")

# stage 3
conv3 = mx.symbol.Convolution(data=pool2, kernel=(3, 3),num_filter=60,workspace=1024)
pool3 = mx.symbol.Pooling(data=conv3, kernel=(2, 2), stride=(2, 2), pool_type="max")

# stage 4
conv4 = mx.symbol.Convolution(data=pool3, kernel=(2, 2),stride=(2, 2), num_filter=80,workspace=1024)

# stage 51
flatten_conv4 = mx.symbol.Flatten(conv4)

# stage 52
flatten_pool3 = mx.symbol.Flatten(pool3)

# stage 6
concat6=mx.symbol.Concat(*[flatten_conv4,flatten_pool3])

# stage 7
fc1 = mx.symbol.FullyConnected(data=concat6, num_hidden=160)
bn1 = mx.symbol.BatchNorm(data=fc1)

# stage 8
#dropout1 = mx.symbol.Dropout(data=fc1, p=0.5)

# stage 9
#fc2 = mx.symbol.FullyConnected(data=dropout1, num_hidden=15529)
fc2 = mx.symbol.FullyConnected(data=bn1, num_hidden=15529)
#bn2 = mx.symbol.BatchNorm(data=fc2)

# stage 9
softmax = mx.symbol.SoftmaxOutput(data=fc2, name='softmax')
#softmax = mx.symbol.SoftmaxOutput(data=bn2, name='softmax')
arg_names = softmax.list_arguments()
print arg_names

initDict={'convolution0_weight':{'normal':0.125}, 'convolution0_bias':{'costant':0}, \
		'convolution1_weight':{'normal':0.2}, 'convolution1_bias':{'costant':0.1}, \
                'convolution2_weight':{'normal':0.2}, 'convolution2_bias':{'costant':0}, \
                'convolution3_weight':{'normal':0.5}, 'convolution3_bias':{'costant':0.1}, \
                'fullyconnected0_weight':{'normal':0.0001}, 'fullyconnected0_bias':{'costant':0.1}, \
                'fullyconnected1_weight':{'xavier':("uniform", "avg", 3)}, 'fullyconnected1_bias':{'costant':0}} 
## data
batch_size = 256 
train, val = deepid_iterator(batch_size=batch_size, input_shape=(1,100,100))

lr = mx.lr_scheduler.FactorScheduler(step=10000, factor=.9)
optimizer = mx.optimizer.SGD(
learning_rate = 0.001,
momentum = 0.9,
wd = 0.005,
lr_scheduler = lr,
clip_gradient = 10)

## train
num_gpus = 2 
gpus = [mx.gpu(i) for i in range(num_gpus)]

model = mx.model.FeedForward(
    ctx           = mx.gpu(0),
    symbol        = softmax,
    num_epoch     = 20,
    initializer   = mx.initializer.Custom(idict=initDict),
    optimizer=optimizer)
    
"""
model = mx.model.FeedForward(
    ctx           = gpus,
    symbol        = softmax,
    num_epoch     = 20,
    learning_rate = 0.0001,
    momentum      = 0.9,
    initializer   = mx.initializer.Custom(idict=initDict),
    wd            = 0.00001)
"""

logging.basicConfig(level = logging.DEBUG)

save_model_prefix = './model/deepid'+prestr
checkpoint = mx.callback.do_checkpoint(save_model_prefix)

model.fit(X = train, eval_data = val,
          batch_end_callback = mx.callback.Speedometer(batch_size=batch_size),
          epoch_end_callback = checkpoint)
```

**Comment by piiswrong** [comment]
are you sure this the the same parameter as caffe solver?
optimizer = mx.optimizer.SGD(
learning_rate = 0.001,
momentum = 0.9,
wd = 0.005,
lr_scheduler = lr,
clip_gradient = 10)

note that wd can have different meaning in caffe than in mxnet. Try set it to 0 first. Then try tuning the learning rate bigger/smaller.

Does caffe deepid use clip_gradient? this can also have different meaning.

**Comment by dushoufu** [comment]
Not used, some parameters come from examples from mxnet examples.

#############################  DATA Layer  #############################
name: "deepID_train_test"
layer {
  name: "deepID"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  transform_param {
    scale: 0.00390625
    mean_file: "/work/chenwj/caffe/caffe-master-160104/caffe/examples/deepID/crop_press.binaryproto"
    mirror:true
  }
  data_param {
    source: "/work_chenwj/chenwj/caffe/caffe-master-160104/caffe/examples/deepID/asia_train_lmdb_100_100_gray_crop_press"
    batch_size: 256
    backend: LMDB
  }
}
layer {
  name: "deepID"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  transform_param {
    scale: 0.00390625
    mean_file: "/work/chenwj/caffe/caffe-master-160104/caffe/examples/deepID/casia_crop_press.binaryproto"
    mirror:false
  }
  data_param {
    source: "/work_chenwj/chenwj/caffe/caffe-master-160104/caffe/examples/deepID/casia_val_lmdb_100_100_gray_crop_press"
    batch_size: 50
    backend: LMDB
  }
}

#############################  CONV NET 1 #############################
layer {
  name: "conv1_1"
  type: "Convolution"
  bottom: "data"
  top: "conv1_1"
  param {
    name: "conv1_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv1_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 20
    kernel_size: 4
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.125
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu1_1"
  type: "ReLU"
  bottom: "conv1_1"
  top: "conv1_1"
}
layer {
  name: "pool1_1"
  type:  "Pooling"
  bottom: "conv1_1"
  top: "pool1_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv2_1"
  type: "Convolution"
  bottom: "pool1_1"
  top: "conv2_1"
  param {
    name: "conv2_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv2_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 40
    kernel_size: 3
    group: 2
    weight_filler {
      type: "gaussian"
      std: 0.2
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }

}
layer {
  name: "relu2_1"
  type: "ReLU"
  bottom: "conv2_1"
  top: "conv2_1"
}
layer {
  name: "pool2_1"
  type:  "Pooling"
  bottom: "conv2_1"
  top: "pool2_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv3_1"
  type: "Convolution"
  bottom: "pool2_1"
  top: "conv3_1"
  param {
    name: "conv3_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv3_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 60
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.2
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }

}
layer {
  name: "pool3_1"
  type:  "Pooling"
  bottom: "conv3_1"
  top: "pool3_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv4_1"
  type: "Convolution"
  bottom: "pool3_1"
  top: "conv4_1"
  param {
    name: "conv4_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv4_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 80
    kernel_size: 2
    stride: 2
    weight_filler {
      type: "gaussian"
      std: 0.5
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }

}
layer{
  name:"flatten_pool3_1"
  type:"Flatten"
  bottom:"pool3_1"
  top:"flatten_pool3_1"
}
layer{
  name:"flatten_conv4_1"
  type:"Flatten"
  bottom:"conv4_1"
  top:"flatten_conv4_1"
}
layer{
  name:"contact_conv"
  type:"Concat"
  bottom:"flatten_conv4_1"
  bottom:"flatten_pool3_1"
  top:"contact_conv"
}
layer {
  name: "deepid_1"
  type:  "InnerProduct"
  bottom: "contact_conv"
  top: "deepid_1"
  param {
    name: "fc6_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "fc6_b"
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 160
    weight_filler {
      type: "gaussian"
      std: 0.0001
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "drop"
  type: "Dropout"
  bottom: "deepid_1"
  top: "deepid_1"
  dropout_param {
    dropout_ratio: 0.5 
  }
}
layer {
  name: "deepid_2"
  type: "InnerProduct"
  bottom: "deepid_1"
  top: "deepid_2"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 15519
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "deepid_2"
  bottom: "label"
  top: "accuracy"
  include {
    phase: TEST
  }
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "deepid_2"
  bottom: "label"
  top: "loss"
}

**Comment by dushoufu** [comment]
I am trying your suggestion and the result is amasing and desirable. I am grateful your help. Most time,I find it is difficult to learn mxnet. Now i learn mxnet by step into.

**Comment by haria** [comment]
@dushoufu why don't you convert your caffe's model and prototxt to mxnet's params and symbol and have a test, then training and tuning based on the converted symbol? I think this is a better way than construct everything from scratch.

**Comment by dushoufu** [comment]
I ever used convert_model and convert_symbol,but there were a lot of errors I needed to deal with, I think it is easy to use mxnet to write a new one.

**Comment by haria** [comment]
@dushoufu decrease lr to some value like 0.0005 maybe helpful, I encounter some problems when try vgg16 on mxnet, but when decrease lr, it converged.

**Comment by dushoufu** [comment]
Mine converged normally, the main problem is that the train accuracy is too low only at like 40%. But with the suggestion of @piiswrong. At the moment it seems to be much better. I am waiting the training result now.

**Comment by dushoufu** [comment]
I changed the parameters as @piiswrong suggested, It reached 54%, making much progress than before. I'm trying tuning the lr to increase the train-accuracy further. But how to set weight_decay in mxnet?

**Comment by qiaohaijun** [comment]
"""
Reference:
xxxx
xxxx
"""

import find_mxnet
import mxnet as mx

def get_symbol(num_classes = 530):
    input_data = mx.symbol.Variable(name="data")
    # stage 1
    conv1 = mx.symbol.Convolution(
        data=input_data, kernel=(4, 4), stride=(1, 1), num_filter=20)
    bn1 = mx.symbol.BatchNorm(data=conv1, name='bn1')

    relu1 = mx.symbol.Activation(data=bn1, act_type="relu")
    pool1 = mx.symbol.Pooling(
        data=relu1, pool_type="max", kernel=(2, 2), stride=(2,2))
    
    # stage 2
    conv2 = mx.symbol.Convolution(
        data=pool1, kernel=(3, 3), stride=(1, 1), num_filter=40)
    bn2 = mx.symbol.BatchNorm(data=conv2, name='bn2')
    relu2 = mx.symbol.Activation(data=bn2, act_type="relu")
    pool2 = mx.symbol.Pooling(data=relu2, kernel=(2, 2), stride=(1, 1), pool_type="max")
    
    # stage 3
    conv3 = mx.symbol.Convolution(
        data=pool2, kernel=(3, 3), stride=(1, 1), num_filter=60)
    bn3 = mx.symbol.BatchNorm(data=conv3, name='bn3')
    relu3 = mx.symbol.Activation(data=bn3, act_type="relu")
    pool3 = mx.symbol.Pooling(data=relu3, kernel=(2, 2), stride=(2, 2), pool_type="max")


    conv4 = mx.symbol.Convolution(
        data=pool3, kernel=(2, 2), stride=(1, 1), num_filter=80)
    bn4 = mx.symbol.BatchNorm(data=conv4, name='bn4')
    relu4 = mx.symbol.Activation(data=conv4, act_type="relu")

    flatten_pool3 = mx.symbol.Flatten(data=pool3)
    flatten_relu4 = mx.symbol.Flatten(data=relu4)

    # stage 4
    fc160_1 = mx.symbol.FullyConnected(data=flatten_pool3, num_hidden=80)
    fc160_2 = mx.symbol.FullyConnected(data=flatten_relu4, num_hidden=80)
    fc160 = mx.symbol.Concat(*[fc160_1, fc160_2])

    #fc160 = fc160_1 + fc160_2
    #fc160 = mx.symbol.ElementWiseSum( [fc160_1,fc160_2])
	
    dropout = mx.symbol.Dropout(data=fc160, p=0.4)
    fc_class = mx.symbol.FullyConnected(data=dropout, num_hidden=num_classes)
    # stage 6
    softmax = mx.symbol.SoftmaxOutput(data=fc_class, name='softmax')
    return softmax

**Comment by dushoufu** [comment]
HI @qiaohaijun ,could you give me your parameters as below. I use your network and my parameters as below and found the result is not desirable.In fact it seems worse than previous. I'm a lot puzzled.
model = mx.model.FeedForward(
    ctx           = gpus,
    symbol        = softmax,
    num_epoch     = 20,
    learning_rate = 0.0001,
    momentum      = 0.9,
    initializer   = mx.initializer.Normal(sigma=0.01),
    wd            = 0.0001)

**Comment by qiaohaijun** [comment]
did you do some data preprocessing?

**Comment by dushoufu** [comment]
There is not some special preprocessing only about face alignment.

**Comment by dushoufu** [comment]
I understand mine is true and this issue can be closed.

**Comment by zxfzqf** [comment]
@dushoufu  Hi!
I met same problem when training DeepID,  Train-accuracy Between 0.4 and 0.6。How did you solve the problem? The data set I use is webface ....

**Comment by dushoufu** [comment]
For me, this is because the data magnitude is too large, and the network as deepid can fit to such a extent only. And you can test your model on other dataset, perhaps you'll find the model can reach a good result.
